lib/avro_turf/connection_wrapper_with_token_auth.rb

```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+class AvroTurf::ConnectionWrapperWithAuthToken < AvroTurf::ConnectionWrapper
+  REFRESH_TOKEN_TRIES = 4
+
+  def initialize(url,
+                 logger:,
+                 proxy: nil,
+                 user: nil,
+                 password: nil,
+                 ssl_ca_file: nil,
+                 client_cert: nil,
+                 client_key: nil,
+                 client_key_pass: nil,
+                 client_cert_data: nil,
+                 client_key_data: nil,
+                 oauth_url: nil,
+                 oauth_client_id: nil,
+                 oauth_client_secret: nil)
+    @oauth_url = oauth_url
+    @oauth_client_id = oauth_client_id
+    @oauth_client_secret = oauth_client_secret
+    @semaphore = Mutex.new
+    @logger = logger
+    @refresh_token_retries_remaining = REFRESH_TOKEN_TRIES
+    refresh_token
+
+    super(url,
+      logger: logger,
+      proxy: proxy,
+      user: user,
+      password: password,
+      ssl_ca_file: ssl_ca_file,
+      client_cert: client_cert,
+      client_key: client_key,
+      client_key_pass: client_key_pass,
+      client_cert_data: client_cert_data,
+      client_key_data: client_key_data)
+  end
+
+  def with_connection
+    result = super
+    @refresh_token_retries_remaining = REFRESH_TOKEN_TRIES
+    result
+  rescue Excon::Error::Unauthorized
```

Comment:
Can't we call refresh before the request (it's just a if if refresh is not needed, right?)? I think we can reduce requests this way?